File: search/binary.c
// binary.c
#include <stdio.h>


int binary(int *V, int N, int elem) {
    int start, middle, end;
    start = 0;
    end = N - 1;
    while (start <= end) {
        middle = (start + end) / 2;
        if (elem < V[middle]) {
            end = middle - 1;
        } else {
            if (elem > V[middle]) {
                start = middle + 1;
            } else {
                return middle;
            }
        }
    }
    return -1;
}

int main() {
    int v[10] = {-8, -5, 1, 4, 14, 21, 23, 54, 57, 90};
    if (binary(v, 10, 54) != -1)
        printf("OK!\n");
    else
        printf("NOT FOUND!\n");
    return 1;
}

File: search/binary.py

def dynamic(l, n):
    start, end = 0, len(l) - 1
    while start <= end:
        middle = (start + end) // 2
        if l[middle] > n:
            end = middle - 1
        elif l[middle] < n:
            start = middle + 1
        else:
            return middle
    return -1


def recursive(l, n, start=0, end=0):
    if start > end:
        return -1

    end = end if end > 0 else len(l) - 1
    middle = (start + end) // 2

    if l[middle] > n:
        end = middle - 1
        return recursive(l, n, start=start, end=end)
    elif l[middle] < n:
        start = middle + 1
        return recursive(l, n, start=start, end=end)
    else:
        return middle


if __name__ == "__main__":
    l = [1, 2, 3, 4, 5, 6, 7, 8 , 9, 10]
    assert dynamic(l, 7) == 6
    assert dynamic(l, 50) == -1
    assert recursive([1, 2, 3, 4], 3) == 2
    assert recursive(l, 50) == -1
